libminifi/src/utils/StringUtils.cpp

```diff
@@ -0,0 +1,162 @@
+  return trimRight(trimLeft(s));
+}
+
+std::vector<std::string> StringUtils::split(const std::string &str, const std::string &delimiter) {
+  std::vector<std::string> result;
+  auto curr = str.begin();
+  auto end = str.end();
+  auto is_func = [delimiter](int s) {
+    return delimiter.at(0) == s;
+  };
+  while (curr != end) {
+    curr = std::find_if_not(curr, end, is_func);
+    if (curr == end) {
+      break;
+    }
+    auto next = std::find_if(curr, end, is_func);
+    result.push_back(std::string(curr, next));
+    curr = next;
+  }
+
+  return result;
+}
+
+bool StringUtils::StringToFloat(std::string input, float &output, FailurePolicy cp /*= RETURN*/) {
+  try {
+    output = std::stof(input);
+  } catch (const std::invalid_argument &ie) {
+    switch (cp) {
+      case RETURN:
+      case NOTHING:
+        return false;
+      case EXIT:
+        exit(1);
+      case EXCEPT:
+        throw ie;
+    }
+  } catch (const std::out_of_range &ofr) {
+    switch (cp) {
+      case RETURN:
+      case NOTHING:
+        return false;
+      case EXIT:
+        exit(1);
+      case EXCEPT:
+        throw ofr;
+    }
+  }
+
+  return true;
+}
+
+std::string StringUtils::replaceEnvironmentVariables(std::string& original_string) {
+  int32_t beg_seq = 0;
+  int32_t end_seq = 0;
+  std::string source_string = original_string;
+  do {
+    beg_seq = source_string.find("${", beg_seq);
+    if (beg_seq > 0 && source_string.at(beg_seq - 1) == '\\') {
+      beg_seq += 2;
+      continue;
+    }
+    if (beg_seq < 0)
+      break;
+    end_seq = source_string.find("}", beg_seq + 2);
+    if (end_seq < 0)
+      break;
+    if (end_seq - (beg_seq + 2) < 0) {
+      beg_seq += 2;
+      continue;
+    }
+    const std::string env_field = source_string.substr(beg_seq + 2, end_seq - (beg_seq + 2));
+    const std::string env_field_wrapped = source_string.substr(beg_seq, end_seq + 1);
+    if (env_field.empty()) {
+      continue;
+    }
+    const auto strVal = std::getenv(env_field.c_str());
+    std::string env_value;
+    if (strVal != nullptr)
+      env_value = strVal;
+    source_string = replaceAll(source_string, env_field_wrapped, env_value);
+    beg_seq = 0;  // restart
+  } while (beg_seq >= 0);
+
+  source_string = replaceAll(source_string, "\\$", "$");
+
+  return source_string;
+}
+
+std::string& StringUtils::replaceAll(std::string& source_string, const std::string &from_string, const std::string &to_string) {
+  std::size_t loc = 0;
+  std::size_t lastFound;
+  while ((lastFound = source_string.find(from_string, loc)) != std::string::npos) {
+    source_string.replace(lastFound, from_string.size(), to_string);
+    loc = lastFound + to_string.size();
+  }
+  return source_string;
+}
+
+std::string StringUtils::replaceMap(std::string source_string, const std::map<std::string, std::string> &replace_map) {
+  auto result_string = source_string;
+
+  std::vector<std::pair<size_t, std::pair<size_t, std::string>>> replacements;
```

Comment:
this can be shortened and improved. I know you didn't do it ( for all I know I could have done it ), but seeing it here highlights that so thanks!